bot/cogs/news.py

```diff
@@ -0,0 +1,182 @@
+import asyncio
+import datetime
+import json
+import logging
+import os
+
+import discord
+from discord.ext import commands, tasks
+
+from bot.bot import WarnetBot
+from bot.cogs.ext.news.genshin import get_genshin_news
+from bot.cogs.ext.news.hoyolab import hoyolab_news
+from bot.config import news as news_config
+
+logger = logging.getLogger(__name__)
+
+
+class News(commands.GroupCog):
+    def __init__(self, bot: WarnetBot) -> None:
+        self.bot = bot
+
+    @commands.Cog.listener()
+    async def on_connect(self) -> None:
+        if not os.path.exists(news_config.LAST_ID_GENSHIN_NEWS_PATH):
+            with open(news_config.LAST_ID_GENSHIN_NEWS_PATH, "w", encoding="utf-8") as f:
+                f.write("")
+
+        if not os.path.exists(news_config.LAST_ID_HOYOLAB_NEWS_PATH):
+            with open(news_config.LAST_ID_HOYOLAB_NEWS_PATH, "w", encoding="utf-8") as f:
+                f.write("")
+
+        if not self._news_hoyolab.is_running():
+            self._news_hoyolab.start()
+        if not self._news_genshin.is_running():
+            self._news_genshin.start()
+
+    @tasks.loop(time=news_config.TIMES_CHECK_UPDATE)
+    async def _news_genshin(self) -> None:
+        info_channel = self.bot.get_channel(news_config.INFORMATION_CHANNEL_ID)
+
+        try:
+            await get_genshin_news()
+        except Exception as e:
+            logger.error(f"[genshin] Failed to fetch Genshin news: {e}")
+            return
+
+        with open(news_config.LAST_ID_GENSHIN_NEWS_PATH, "r", encoding="utf-8") as f:
+            last_id = f.read().strip()
+
+        with open(news_config.JSON_GENSHIN_NEWS_PATH, "r", encoding="utf-8") as f:
+            news_data = json.load(f)["data"]["list"]
+
+        last_json_id = str(news_data[0]["iInfoId"])
+        if last_json_id == last_id:
+            logger.info("[genshin] No new news updates found.")
+            return
+
+        start_index = None
+        for i, item in enumerate(news_data):
+            if str(item["iInfoId"]) == last_id:
+                start_index = i
+                break
+
+        if start_index is None:
+            new_news = news_data
+        else:
+            new_news = news_data[:start_index]
+
+        tag_color_map = {
+            "Info": discord.Color.blue(),
+            "Events": discord.Color.green(),
+            "Updates": discord.Color.gold(),
+        }
+        chan_id_map = {
+            "396": "Info",
+            "397": "Updates",
+            "398": "Events",
+        }
```

Comment:
Better put these as class attribute for the sake of performance and memory saving. Initialize the vars in `__init__`